Source/qol/stash.cpp

```diff
@@ -607,4 +612,51 @@ void GoldWithdrawNewText(string_view text)
 	}
 }
 
+bool AutoPlaceItemInStash(Player &player, const Item &item, bool persistItem)
+{
+	if (item._itype == ItemType::Gold) {
+		if (persistItem) {
+			Stash.gold += item._ivalue;
+			Stash.dirty = true;
+		}
+		return true;
+	}
+
+	Size itemSize = GetInventorySize(item);
+
+	constexpr int CountStashPages = 50;
+
+	// Try to add the item to the current active page and if it's not possible move forward
+	for (int pageCounter = 0; pageCounter < CountStashPages; pageCounter++) {
+		int pageIndex = Stash.page + pageCounter;
+		// Wrap around if needed
+		if (pageIndex >= CountStashPages)
+			pageIndex -= CountStashPages;
+		// Search all possible position in stash grid
+		for (auto stashPosition : PointsInRectangleRange({ { 0, 0 }, { 10 - (itemSize.width - 1), 10 - (itemSize.height - 1) } })) {
+			// Check that all needed slots are free
+			bool isSpaceFree = true;
+			for (auto itemPoint : PointsInRectangleRange({ { 0, 0 }, itemSize })) {
+				uint16_t iv = Stash.stashGrids[pageIndex][stashPosition.x + itemPoint.x][stashPosition.y + itemPoint.y];
+				if (iv != 0) {
+					isSpaceFree = false;
+					break;
+				}
+			}
+			if (!isSpaceFree)
+				continue;

```

Comment:
Some of this could probably be shared with other parts of the code.

Reply:
I thought about reusing it in `CheckStashPaste`.
But `CheckStashPaste` works differently. if we find exactly one item at this space this item is swapped with the pasted item.
This isn't behavior isn't desired in `AutoPlaceItemInStash`.
That's why I didn't combine them. After that I didn't investigate further.